colossalai/shardformer/policies/opt.py

```diff
@@ -114,10 +183,29 @@ def module_policy(self):
                 suffix="lm_head", target_module=Linear1D_Col, kwargs=dict(gather_output=True)),
                                                         policy=policy,
                                                         target_key=OPTForCausalLM)
+
+        if self.pipeline_stage_manager:
+            self.set_pipeline_forward(model_cls=OPTForCausalLM,
+                                      new_forward=OPTPipelineForwards.opt_for_causal_lm_forward,
+                                      policy=policy)
+
         return policy
 
+    def get_held_layers(self) -> List[nn.Module]:
+        held_layers = super().get_held_layers()
+        if self.pipeline_stage_manager.is_last_stage():
+            held_layers.append(self.model.lm_head)
+        return held_layers
+
+    def get_shared_params(self) -> List[Dict[int, Tensor]]:
+        opt_model = self.model
+        if self.pipeline_stage_manager:
+            num_stages = self.pipeline_stage_manager.num_stages
+            if id(opt_model.model.decoder.embed_tokens.weight) == id(opt_model.lm_head.weight):
```

Comment:
Check whether these params are in the different stages here.

Reply:
done